grafana/grafana
# Query History: Refactor persistence layer

**What this PR does / why we need it**:

As part of saving Rich History in the backend, we will need to support two persistence layers at the same time (local storage and backend storage), controlled with a feature flag for testing.

The current implementation is heavily based on local storage. This PR extracts basic persistence actions so they can be easily replaced with backend implementation.

**Which issue(s) this PR fixes**:



Fixes #44969

**Special notes for your reviewer**:

There should be no change in behavior. This PR refactors just basic API for getting/storing items. Following PRs will refactor filtering and settings.

How to test it?
* Check if your query history loads correctly when switching to this branch.
* Add a new query and verify it was added.
* Add the same query once again - duplicates should not be added.
* Try starring, commenting, and deleting. Make sure that the list of queries is the same after refreshing.
* Ensure filtering, searching and settings work as in main branch.
* To test local storage quota errors:

In RichHistoryLocalStorage.addToRichHistory change try/catch block to:

```typescript
    try {
      const e = new Error('Quota exceeded');
      e.name = 'QuotaExceededError';
      throw e;
      store.setObject(RICH_HISTORY_KEY, updatedHistory);
    }
```

There should be an error shown in the top right corner but only once per session.

* To test max limits in the same file change MAX_HISTORY_ITEMS:

```typescript
export const MAX_HISTORY_ITEMS = 5;
```

Try to add new items. A warning message should be shown but only once per session.
